# Task 1 and 2: Pokemon EDA

In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Task 1: Setting the Aesthetics for the plots

### 1.1: Set the Seaborn figure theme and scale up the text in the figures

There are five preset Seaborn themes: `darkgrid`, `whitegrid`, `dark`, `white`, and `ticks`. 
They are each suited to different applications and personal preferences.
You can see what they look like [here](https://seaborn.pydata.org/tutorial/aesthetics.html#seaborn-figure-styles).

Hint: You will need to use the `font_scale` property of the `set_theme()` function in Seaborn.

In [18]:
# Your Solution Here
sns.set_style("white")
sns.set_style("ticks")
sns.set_theme(font_scale = 1.5)

Once you are able to set the theme, you will see all plots in this Jupyter Notebook update using the same theme.

Remember to copy this code above to your other Jupyter Notebooks as well!! 

## Task 2: Exploratory Data Analysis 

In this section of the task, you will be performing EDA on a given dataset with a goal to be able to describe it. 

### 2.1. Describe your dataset 

Consider (and keep in mind) the following questions to guide you in your exploration:

- Who: Which company/agency/organization provided this data?
- What: What is in your data?
- When: When was your data collected (for example, for which years)?
- Why: What is the purpose of your dataset? Is it for transparency/accountability, public interest, fun, learning, etc...
- How: How was your data collected? Was it a human collecting the data? Historical records digitized? Server logs?

**Hint:** The [pokemon dataset is from this Kaggle page.](https://www.kaggle.com/rounakbanik/pokemon)


**Hint:** *You probably will not need more than 250 words to describe your dataset. All the questions above do not need to be answered, it's more to guide your exploration and think a little bit about the context of your data. It is also possible you will not know the answers to some of the questions above, that is FINE - data scientists are often faced with the challenge of analyzing data from unknown sources. Do your best, acknowledge the limitations of your data as well as your understanding of it. Also, make it clear what you're speculating about. For example, "I speculate that the {...column_name...} column must be related to {....} because {....}."*


# Your Dataset Description here (convert this to a markdown cell and write some text)

The dataset in hand contains information on 802 Pokemon from the all Seven Generations of Pokemon. Some of the provided information are: name, height, weight, type, abilities, base stats, and more. The original creator, Rounak Banik, credits his love of the game from his early days up to his adulthood as a reason to gather the data.

The data is gathered from https://serebii.net/ and is compiled into a dataframe in https://www.kaggle.com/datasets/rounakbanik/pokemon.

### 2.2. Load the dataset from a file, or URL 

This needs to be a pandas dataframe. Remember that others may be running your jupyter notebook so it's important that the data is accessible to them. 

***General advice:** If your dataset isn't accessible as a URL, make sure to commit it into your repo. If your dataset is too large to commit (>100 MB), and it's not possible to get a URL to it, you should contact your instructor for advice.*

For this question, luckily the dataset is available as a URL.
You can use this URL to load the data: https://github.com/firasm/bits/raw/master/pokemon.csv

In [36]:
# Your solution here
df = pd.read_csv("https://github.com/firasm/bits/raw/master/pokemon.csv")
df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

### 2.3. Explore your dataset 

Which of your columns are interesting/relevant? Remember to take some notes on your observations, you'll need them for the next EDA step (initial thoughts).

#### 2.3.1:  You should start with [`df.describe().T`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) 

See [linked documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) for the use of `include`/`exclude` to look at numerical and categorical data.

In [4]:
# Your solution to output `df.describe.T` for numerical columns:
df.describe(include = [np.number]).T

count       mean         std    min     25%    50%     75%    max
#           800.0  362.81375  208.343798    1.0  184.75  364.5  539.25  721.0
Total       800.0  435.10250  119.963040  180.0  330.00  450.0  515.00  780.0
HP          800.0   69.25875   25.534669    1.0   50.00   65.0   80.00  255.0
Attack      800.0   79.00125   32.457366    5.0   55.00   75.0  100.00  190.0
Defense     800.0   73.84250   31.183501    5.0   50.00   70.0   90.00  230.0
Sp. Atk     800.0   72.82000   32.722294   10.0   49.75   65.0   95.00  194.0
Sp. Def     800.0   71.90250   27.828916   20.0   50.00   70.0   90.00  230.0
Speed       800.0   68.27750   29.060474    5.0   45.00   65.0   90.00  180.0
Generation  800.0    3.32375    1.661290    1.0    2.00    3.0    5.00    6.0

In [5]:
# Your solution to output `df.describe.T` for categorical columns:
df.describe(exclude = [np.number]).T

count unique        top freq
Name        800    800  Bulbasaur    1
Type 1      800     18      Water  112
Type 2      414     18     Flying   97
Legendary   800      2      False  735

### 2.4. Initial Thoughts

#### 2.4.1. Use this section to record your observations. 

Does anything jump out at you as surprising or particularly interesting? Feel free to make additional plots as needed to explore your data set.

Where do you think you'll go with exploring this dataset? Feel free to take notes in this section and use it as a scratch pad.

Any content in this area will only be marked for effort and completeness.

#### # Your observations here:

- Taking the max amount of each stat, the best pokemon will have: 255 HP, 190 Attack, 230 Defense, and 180 Speed.
- One path one can take would be who is the best value pokemon in terms of the most balanced between all stats.
- Another path could be who has the best Attack to Defense balance ratio.
- Does speed affect attack and defense stats?

### 2.5. Wrangling 

The next step is to wrangle your data based on your initial explorations. Normally, by this point, you have some idea of what your research question will be, and that will help you narrow and focus your dataset. 

In this lab, we will guide you through some wrangling tasks with this dataset.

#### 2.5.1. Drop the 'Generation', 'Sp. Atk', 'Sp. Def', 'Total', and the '#' columns

In [37]:
# Your solution here
df2 = df.drop(['Generation', 'Sp. Atk','Sp. Def','Total','#'], axis = 1)
df2

Name   Type 1  Type 2  HP  Attack  Defense  Speed  \
0                Bulbasaur    Grass  Poison  45      49       49     45   
1                  Ivysaur    Grass  Poison  60      62       63     60   
2                 Venusaur    Grass  Poison  80      82       83     80   
3    VenusaurMega Venusaur    Grass  Poison  80     100      123     80   
4               Charmander     Fire     NaN  39      52       43     65   
..                     ...      ...     ...  ..     ...      ...    ...   
795                Diancie     Rock   Fairy  50     100      150     50   
796    DiancieMega Diancie     Rock   Fairy  50     160      110    110   
797    HoopaHoopa Confined  Psychic   Ghost  80     110       60     70   
798     HoopaHoopa Unbound  Psychic    Dark  80     160       60     80   
799              Volcanion     Fire   Water  80     110      120     70   

     Legendary  
0        False  
1        False  
2        False  
3        False  
4        False  
..         ...  
795       True  
796       True  
797       True  
798       True  
799       True  

[800 rows x 8 columns]

#### 2.5.2. Drop any NaN values in HP, Attack, Defense, Speed

In [39]:
# Your solution here
df2 = df2.dropna(subset=['HP', 'Attack', 'Defense', 'Speed'], inplace=True)
df2

AttributeError: 'NoneType' object has no attribute 'dropna'

#### 2.5.3. Reset the index to get a new index without missing values

In [34]:
# Your solution here
df = df.reset_index()
df

AttributeError: 'NoneType' object has no attribute 'reset_index'

#### 2.5.4. A new column was added called `index`; remove it. 

In [35]:
# Your solution here
df = df.drop(['index'], axis = 1)
df

AttributeError: 'NoneType' object has no attribute 'drop'

#### 2.5.5. Calculate a new column called "Weighted Score" that computes an aggregate score comprising:

- 20% 'HP'
- 40% 'Attack'
- 30% 'Defense'
- 10% 'Speed'


In [73]:
# Your solution here
df['Weighted Score'] = (df['HP'] * 0.2) + (df['Attack'] * 0.4) + (df['Defense'] * 0.3) + (df['Speed'] * 0.1)
df

Name   Type 1  Type 2  HP  Attack  Defense  Speed  \
0                Bulbasaur    Grass  Poison  45      49       49     45   
1                  Ivysaur    Grass  Poison  60      62       63     60   
2                 Venusaur    Grass  Poison  80      82       83     80   
3    VenusaurMega Venusaur    Grass  Poison  80     100      123     80   
4                Charizard     Fire  Flying  78      84       78    100   
..                     ...      ...     ...  ..     ...      ...    ...   
409                Diancie     Rock   Fairy  50     100      150     50   
410    DiancieMega Diancie     Rock   Fairy  50     160      110    110   
411    HoopaHoopa Confined  Psychic   Ghost  80     110       60     70   
412     HoopaHoopa Unbound  Psychic    Dark  80     160       60     80   
413              Volcanion     Fire   Water  80     110      120     70   

     Legendary  Weighted Score  
0        False            47.8  
1        False            61.7  
2        False            81.7  
3        False           100.9  
4        False            82.6  
..         ...             ...  
409       True           100.0  
410       True           118.0  
411       True            85.0  
412       True           106.0  
413       True           103.0  

[414 rows x 9 columns]

### 2.6. Research questions 

#### 2.6.1 Come up with at least two research questions about your dataset that will require data visualizations to help answer. 

Recall that for this purpose, you should only aim for "Descriptive" or "Exploratory" research questions.

**Hint1: You are welcome to calculate any columns that you think might be useful to answer the question (or re-add dropped columns like 'Generation', 'Sp. Atk', 'Sp. Def'.***

**Hint2: Try not to overthink this; this is a toy dataset about Pokémon, you're not going to solve climate change or cure world hunger. Focus your research questions on the various Pokémon attributes, and the types.**

#### # Your solution here: 

**1. Sample Research Question:** Which Pokemon Types are the best, as determined by the Weighted Score?

**2. Your RQ 1: Which Pokemon itself is the best, as determined by the Weighted Score?

**3. Your RQ 2: While speed only accounts for 10% of the Weighted Score, does it have a lot of correlation in terms of the total score regardless? 



### 2.7. Save your dataset

Here, using the [pandas.DataFrame.to_csv](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) function, save your dataframe to be used in other tasks by naming it **task2.csv** in the data directory.

In [74]:
# Your solution here. 
df.to_csv('task2.csv')